src/conv-core/convcore.C

```diff
@@ -1719,7 +1721,21 @@ void CsdSchedulerState_new(CsdSchedulerState_t *s)
  */
 void *CsdNextMessage(CsdSchedulerState_t *s) {
 	void *msg;
-	if((*(s->localCounter))-- >0)
+
+	s->iter++;
+
+#if CMK_NODE_QUEUE_AVAILABLE
+	if (1 == (s->iter & (1 << s->nodeGrpFreq))) // since we use nodeGrpFreq == 0 to mean
```

Comment:
Do we gain from avoiding the modulo operator this way? If the definition of ```s->nodeGrpFreq``` is compile-time constant, can we assume that the compiler will do the corresponding strength reduction, provided that ```s->nodeGrpFreq``` is a power of 2? I checked a recent version of GCC and it generates the same code if we use modulo or this trick

Reply:
Separately, this comment/implementation seems wrong, if `nodeGrpFreq = 0`, then `1 << s->nodeGrpFreq` would be 1, so that would check the node queue every other iteration, seemingly contradicting the comment saying that that would not check the node queue with high priority. Also, testing it via `1 == ` means that it would only work in the case where `nodeGrpFreq = 0`, otherwise it would only ever produce 0 or a value greater than 1.

Reply:
I think you want `if (0 == (s->iter & ((1 << (s->nodeGrpFreq-1))-1)))`
nodeGrpFreq of 0 -> 0 == iter & -1
nodeGrpFreq of 1 -> 0 == iter & 0
nodeGrpFreq of 2 -> 0 == iter & 1
nodeGrpFreq of 3 -> 0 == iter & 3

Reply:
Not sure what 1 << -1 produces.

Reply:
Hoping it's zero.

Reply:
It's not.
`if (s->nodeGrpFreq && ! (s->iter & ((1 << (s->nodeGrpFreq-1))-1)))`

Reply:
Looks like I let this slide.. @ZwFink will you complete it ? You can simplify the code (that I had to avoid the % operator) because I am convinced by your argument that the compiler will optimize it.